bazel/test_runners/qemu_with_kernel/launcher.sh

```diff
@@ -0,0 +1,128 @@
+  retval=$?
+  rm -rf "${tmpdir_for_sandbox}" || true
+  exit "${retval}"
+}
+
+# We set a trap to make sure the tmp directories are cleaned up.
+trap cleanup EXIT
+trap cleanup ERR
+
+function strip_pwd_from_path() {
+  v=$1
+  v=${v//"${OLDPWD}"/""}
+  echo "$v"
+}
+
+# Replaces the PWD outside qemu with /test_fs.
+function path_inside_qemu() {
+  v=$1
+  test_base="/test_fs"
+  v=${v//"${OLDPWD}"/"${test_base}"}
+  echo "$v"
+}
+
+# Replaces the PWD outside qemu with a path to the sandbox directory.
+function path_qemu_sandbox() {
+  v=$1
+  v=${v//"${OLDPWD}"/"${tmpdir_for_sandbox}"}
+  echo "$v"
+}
+
+
+# We need to write and transform the environment variables so that they have the "correct"
+# paths when run inside of qemu.
+test_env_file="${tmpdir_for_sandbox}/test_env.sh"
+echo "#!/bin/bash" > "${test_env_file}"
+
+# Grab the environment variables and rewrite them for the qemu sandbox.
+env -0 | while IFS='=' read -r -d '' n v; do
+  # Writing bash funcs as env variables doesn't work. Luckily we don't actually need them.
+  if [[ "${n}" == BASH_FUNC* ]]; then
+    continue
+  fi
+
+  # We don't want to write out the PWD since it will be wrong inside of qemu.
+  if [[ "${n}" == PWD ]]; then
+    continue
+  fi
+
+  # All paths will basically start with /test_fs.
+  v=$(path_inside_qemu "$v")
+
+  printf "export %s=%s\n" "$n" "$v" >> "${test_env_file}"
+done
+
+
+# Copy the runfiles into the sandbox directory. 9p fs does not like symlinks
+# so we resolve them into the copy.
+runfiles_path="$(path_qemu_sandbox "${RUNFILES_DIR}")"
+mkdir -p "${runfiles_path}"
+cp -aL "${RUNFILES_DIR}"/* "${runfiles_path}"
+
+# Copy over testlog file.
+testlogs_dir="${TEST_WARNINGS_OUTPUT_FILE/${TEST_WARNINGS_OUTPUT_FILE##*/testlogs/}}"
+qemu_warnings_file=$(strip_pwd_from_path "${TEST_WARNINGS_OUTPUT_FILE}")
+qemu_testlogs_dir="${qemu_warnings_file/${qemu_warnings_file##*/testlogs/}}"
+cp -aL "${testlogs_dir}/" "${tmpdir_for_sandbox}/${qemu_testlogs_dir}/"
+
+# Copy the test runner and test cmd into the sandbox.
+test_runner_file="${tmpdir_for_sandbox}/test_runner_inside_qemu.sh"
+test_base=${PWD//"${OLDPWD}"/\/test_fs}
+test_cmd_path="${tmpdir_for_sandbox}/test_cmd.sh"
+test_cmd_path_in_qemu="/test_fs/test_cmd.sh"
+
+echo "#!/bin/bash -e" > "${test_cmd_path}"
+echo "${@:1}" >> "${test_cmd_path}"
+chmod +x "${test_cmd_path}"
+
+printf "export test_base=%s\n" "${test_base}" >> "${test_env_file}"
+printf "export test_exec_path=%s\n" "${test_cmd_path_in_qemu}" >> "${test_env_file}"
+
+cp "${bazel_qemu_test_runner_path}" "${test_runner_file}"
+
+# Launch the qemu test.
+retval=0
+(exec env - \
+  QEMU_TEST_FS_PATH="${tmpdir_for_sandbox}" \
+  QEMU_KERNEL_IMAGE="${KERNEL_IMAGE}" \
+  QEMU_DISK_BASE_RO="${PWD}/${DISK_IMAGE}" \
+  "${run_qemu_script}" ) || retval=$?
+
+# We use a known path to find the testlogs directory so that we can copy the results back from qemu.
+testlogs_dir="${TEST_WARNINGS_OUTPUT_FILE/${TEST_WARNINGS_OUTPUT_FILE##*/testlogs/}}"
```

Comment:
in case there's a second `testlogs/` I think its better to do 
```
"${TEST_WARNINGS_OUTPUT_FILE%%/testlogs/*}/testlogs/"
```
and same below